### PR TITLE
Add SigOpt HPO to transformers trainer api

### PR DESCRIPTION
**Pull request by kding1**
# What does this PR do?
This PR adds SigOpt (https://sigopt.com/) hyper-parameter optimization support to transformer trainer api.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?
@mfuntowicz 

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by mfuntowicz — APPROVED** [review]
LGTM! Thanks @kding1 for enabling SigOpt as another HPO backend 💪🏻

File: src/transformers/integrations.py [hunk 1]
```diff
@@ -288,6 +298,49 @@ def dynamic_modules_import_trainable(*args, **kwargs):
...
+    proxies = kwargs.pop("proxies", None)
+    if proxies is not None:
+        conn.set_proxies(proxies)
+
+    experiment = conn.experiments().create(
+        name="huggingface-tune",
+        parameters=trainer.hp_space(None),
+        metrics=[
+            dict(name="objective", objective=direction, strategy="optimize"),
+        ],
```
Comment (sgugger) [inline src/transformers/integrations.py:315]:
```suggestion
        metrics=[dict(name="objective", objective=direction, strategy="optimize")],
```
This fits in one line

File: src/transformers/integrations.py [hunk 2]
```diff
@@ -288,6 +298,49 @@ def dynamic_modules_import_trainable(*args, **kwargs):
...
+        trainer.objective = None
+        trainer.train(resume_from_checkpoint=None, trial=suggestion)
+        # If there hasn't been any evaluation during the training loop.
+        if getattr(trainer, "objective", None) is None:
+            metrics = trainer.evaluate()
+            trainer.objective = trainer.compute_objective(metrics)
+
+        values = [
+            dict(name="objective", value=trainer.objective),
+        ]
```
Comment (sgugger) [inline src/transformers/integrations.py:333]:
```suggestion
        values = [dict(name="objective", value=trainer.objective)]
```

File: src/transformers/integrations.py [hunk 3]
```diff
@@ -288,6 +298,49 @@ def dynamic_modules_import_trainable(*args, **kwargs):
...
+        obs = conn.experiments(experiment.id).observations().create(suggestion=suggestion.id, values=values)
+        logger.info("[suggestion_id, observation_id]: [%s, %s]" % (suggestion.id, obs.id))
```
Comment (sgugger) [inline src/transformers/integrations.py:335]:
```suggestion
        logger.info(f"[suggestion_id, observation_id]: [{suggestion.id}, {obs.id}]")
```
We only use f-strings in the code base.

File: src/transformers/integrations.py [hunk 4]
```diff
@@ -288,6 +298,49 @@ def dynamic_modules_import_trainable(*args, **kwargs):
...
+        name="huggingface-tune",
+        parameters=trainer.hp_space(None),
+        metrics=[
+            dict(name="objective", objective=direction, strategy="optimize"),
+        ],
+        parallel_bandwidth=1,
+        observation_budget=n_trials,
+        project="huggingface",
+    )
+    logger.info("created experiment: https://app.sigopt.com/experiment/" + experiment.id)
```
Comment (sgugger) [inline src/transformers/integrations.py:320]:
```suggestion
    logger.info(f"created experiment: https://app.sigopt.com/experiment/{experiment.id}")
```

File: src/transformers/trainer.py [hunk 5]
```diff
@@ -883,6 +887,8 @@ def _hp_search_setup(self, trial: Union["optuna.Trial", Dict[str, Any]]):
             setattr(self.args, key, value)
         if self.hp_search_backend == HPSearchBackend.OPTUNA:
             logger.info("Trial:", trial.params)
+        if self.hp_search_backend == HPSearchBackend.SIGOPT:
+            logger.info("SigOpt Assignments: %s" % trial.assignments)
```
Comment (sgugger) [inline src/transformers/trainer.py:891]:
```suggestion
            logger.info(f"SigOpt Assignments: {trial.assignments}")
```

**Review by sgugger — APPROVED** [review]
Thanks a lot for your PR! This is very clean, I only have some suggestions of code formatting.

**Review by LysandreJik — COMMENTED** [review]
I think this looks good! It seems I'm unable to run the test locally, I get the following error:

```
―――――――――――――――――――――――――――――――――――――――――――――――――――――――― TrainerHyperParameterSigOptIntegrationTest.test_hyperparameter_search ―――――――――――――――――――――――――――――――――――――――――――――――――――――――――

self = <tests.test_trainer.TrainerHyperParameterSigOptIntegrationTest testMethod=test_hyperparameter_search>

    def test_hyperparameter_search(self):
        class MyTrialShortNamer(TrialShortNamer):
            DEFAULTS = {"a": 0, "b": 0}

        def hp_space(trial):
            return [
                {"bounds": {"min": -4, "max": 4}, "name": "a", "type": "int"},
                {"bounds": {"min": -4, "max": 4}, "name": "b", "type": "int"},
            ]

        def model_init(trial):
            if trial is not None:
                a = trial.assignments["a"]
                b = trial.assignments["b"]
            else:
                a = 0
                b = 0
            config = RegressionModelConfig(a=a, b=b, double_output=False)

            return RegressionPreTrainedModel(config)

        def hp_name(trial):
            return MyTrialShortNamer.shortname(trial.assignments)

        with tempfile.TemporaryDirectory() as tmp_dir:
            trainer = get_regression_trainer(
                output_dir=tmp_dir,
                learning_rate=0.1,
                logging_steps=1,
                evaluation_strategy=IntervalStrategy.EPOCH,
                save_strategy=IntervalStrategy.EPOCH,
                num_train_epochs=4,
                disable_tqdm=True,
                load_best_model_at_end=True,
                logging_dir="runs",
                run_name="test",
                model_init=model_init,
            )
>           trainer.hyperparameter_search(
                direction="minimize", hp_space=hp_space, hp_name=hp_name, backend="sigopt", n_trials=4
            )

tests/test_trainer.py:1548:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
src/transformers/trainer.py:1746: in hyperparameter_search
    best_run = backend_dict[backend](self, n_trials, direction, **kwargs)
src/transformers/integrations.py:305: in run_hp_search_sigopt
    conn = Connection()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <sigopt.interface.Connection object at 0x7f3fd008aee0>, client_token = None, user_agent = None, session = None

    def __init__(self, client_token=None, user_agent=None, session=None):
      client_token = client_token or os.environ.get('SIGOPT_API_TOKEN', config.api_token)
      api_url = os.environ.get('SIGOPT_API_URL') or DEFAULT_API_URL
      # no-verify overrides a passed in path
      no_verify_ssl_certs = os.environ.get('SIGOPT_API_NO_VERIFY_SSL_CERTS')
      if no_verify_ssl_certs:
        verify_ssl_certs = False
      else:
        verify_ssl_certs = os.environ.get('SIGOPT_API_VERIFY_SSL_CERTS')

      if not client_token:
>       raise ValueError('Must provide client_token or set environment variable SIGOPT_API_TOKEN')
E       ValueError: Must provide client_token or set environment variable SIGOPT_API_TOKEN
```

File: src/transformers/integrations.py (hunk 1 above)
Comment (kding1) [reply src/transformers/integrations.py]: fixed.

File: src/transformers/integrations.py (hunk 2 above)
Comment (kding1) [reply src/transformers/integrations.py]: fixed.

File: src/transformers/integrations.py (hunk 3 above)
Comment (kding1) [reply src/transformers/integrations.py]: fixed.

File: src/transformers/integrations.py (hunk 4 above)
Comment (kding1) [reply src/transformers/integrations.py]: fixed.

File: src/transformers/trainer.py (hunk 5 above)
Comment (kding1) [reply src/transformers/trainer.py]: fixed.

**Comment by kding1** [comment]
@LysandreJik this is because the sigopt token is not set properly in your test environment. Below are the steps:

- signup on sigopt HPO at https://app.sigopt.com/signup, it's free for everyone for the basic plan.
- login to your account, get your sigopt api token at your portal, and expose it in your environment variable with `export SIGOPT_API_TOKEN=your_sigopt_token_string`
- in your account portal, create a project named `huggingface`.

Once the training starts with sigopt HPO, you will get the real-time dashboard like this: https://app.sigopt.com/guest?guest_token=NJTRFVEPPGWTKZPLISXVWUXPIOFATREJOVAILZQBPSLOECMU

I have run the unit test with the refactor patch i just submitted and it works:
`$ pytest tests/test_trainer.py::TrainerHyperParameterSigOptIntegrationTest::test_hyperparameter_search`

Please let me know if there is any issue in your side.

**Comment by kding1** [comment]
rebase to fix conflict on dependency package.

**Pull request merged**